Upgrade Phing to version 3.0.0.

EreMaijala opened the pull request:
This upgrades Phing to version 3.0.0 and modifies the build file accordingly. I've tried to test as much as I can, but there are still several tasks that I cannot test (e.g. Selenium).

EreMaijala committed:
Upgrade Phing to version 3.0.0.

EreMaijala committed:
Fix error when dropping non-existent user.

EreMaijala committed:
Fix another exec error.

EreMaijala committed:
Fix check-bootstrap5

demiankatz commented on composer.json:
```diff
@@ -33,7 +33,7 @@
         "composer/package-versions-deprecated": "1.11.99.5",
         "composer/semver": "3.4.0",
-        "endroid/qr-code": "5.0.9",
+        "endroid/qr-code": "5.0.7",
```
Did you mean to roll this back?

EreMaijala replied:
Certainly not. No idea how I managed to do it without conflicts..

demiankatz commented on build.xml:
```diff
@@ -37,8 +38,8 @@
-  <property name="phpunit_command" value="${srcdir}/vendor/bin/phpunit" /><!-- if you have issues with segfaults during code coverage generation, try adding -dzend.enable_gc=0 to this (but this fix was probably only needed in PHP 5) -->
-  <property name="phpunit_extra_params" value="" />
+  <property name="phpunit_command" value="${srcdir}/vendor/bin/phpunit" />
+  <property name="phpunit_extra_params" value="" /><!-- if you have issues with segfaults during code coverage generation, try adding -dzend.enable_gc=0 to this (but this fix was probably only needed in PHP 5) -->
```
Should we drop the zend.enable_gc comment entirely? I suspect it no longer applies...

EreMaijala replied:
Yes, dropped!

demiankatz commented on build.xml:
```diff
@@ -313,19 +355,34 @@
+    <exec dir="${srcdir}/module/VuFind/tests" executable="${phpunit_command}" passthru="true" checkreturn="true">
+      <arg line="--log-junit ${builddir}/reports/phpunit.xml --coverage-php ${coveragedir}/tmp/unit.cov ${phpunit_extra_params}" />
+      <env key="XDEBUG_MODE" value="coverage" />
+      <env key="VUFIND_MINK_DRIVER" value="${mink_driver}" />
```
Feels like it might be beneficial at some point to refactor so we don't have to copy and paste all these variables to so many different places, but that's a project for another day... just thinking out loud!

EreMaijala replied:
I agree, but it also complicates things. I wish there was support for a set of env vars...

demiankatz commented on build.xml:
```diff
@@ -610,6 +719,11 @@
+        <mkdir dir="${themedir}" />
+        <mkdir dir="${themedir}/templates" />
+        <mkdir dir="${contentdir}" />
+        <mkdir dir="${contentdir}/test" />
```
Would be nice if they had an equivalent to the `-p` switch!

EreMaijala replied:
Agreed. It already works like `-p`, but spits out scary-looking error messages (it did show warnings before as well, but not so scary-looking).

demiankatz commented on build.xml:
```diff
@@ -638,41 +752,86 @@ return [
 
-        <!--<exec command="sudo su -c &quot;psql -c \&quot;select 'grant SELECT,INSERT,UPDATE,DELETE on '||schemaname||'.'||tablename||' to ${vufinddbuser};' from pg_tables where schemaname in ('${vufinddb}') order by schemaname, tablename;\&quot;&quot; ${pgsqlrootuser}" checkreturn="true" />-->
-        <exec command="PGPASSWORD=${vufinddbpass} psql -U ${vufinddbuser} -f ${srcdir}/module/VuFind/sql/pgsql.sql ${vufinddb}" checkreturn="true" />
+        <!--<exec executable="sudo" checkreturn="true">
```
Any idea why this commented-out line is here? Maybe now is a chance to clean up old junk.

EreMaijala replied:
No idea. Now removed.

demiankatz reviewed (changes requested):
Thanks, @EreMaijala, this looks pretty reasonable to me; see below for a few comments.

I haven't actually tried it yet, but I can do so after you make any needed revisions.

I'm not too worried about Selenium -- I doubt it works anymore anyway, and I haven't used it in years. But if there's anything else you'd like me to help test specifically, please let me know!

EreMaijala committed:
Revert bad change.

EreMaijala committed:
Cleanup.

EreMaijala committed:
Use a more modern Composer version by default.

EreMaijala commented:
@demiankatz I've now tested most targets, but `package` is the one that could use more testing.

demiankatz committed:
Merge branch 'dev' into dev-phing3

demiankatz committed:
Fix package task.

demiankatz reviewed (changes requested):
Thanks, @EreMaijala. I tested the `package` task and found that it was broken due to a single missing dollar sign. I've pushed up a fix, and also done a regex search to be sure the error is not repeated anywhere else in the file (it isn't).

Running the task results in comparable results to what I get from Phing2 (though the file sizes are slightly different because Phing itself is part of the bundle, and the new version is a different size than the old version).

I think this is probably safe to merge, but I'm going to hold off on hitting the merge button in case there's anything else you'd like to try first. But if you're comfortable, please go ahead and merge at your convenience, and we'll see if Jenkins turns up any new problems...

(And then there's all the subprojects to deal with... *sigh*).

demiankatz reviewed (approved):
Sorry, meant to approve, not request changes. Force of habit!

demiankatz reviewed (changes requested):
Hmm, maybe we're not ready yet. When I finished up work for the morning, I ran the `shutdown` task and got this:

```
vufind > shutdown:

   [delete] Could not find file /etc/apache2/conf-enabled/vufindtest.conf to delete.
     [exec] Result: 127
     [exec] Result: 1
     [exec] Result: 1
No process found for Solr node running on port 8983
     [exec] Result: 1
     [exec] Result: 127
   [delete] Deleting directory /usr/local/vufind/solr
   [delete] Deleting directory /usr/local/vufind/vendor
   [delete] Deleting directory /usr/local/vufind/local

BUILD FAILED
Class "Phing\Io\IOException" not found
```

Are you seeing the same thing?

Maybe the problem occurs if you run it twice in a row, or if some parts of it aren't necessary -- e.g. shutting down after a failed partial startup. It's possible I double shut-down by accident. But in any case, it left things in a scrambled state, which is not what usually happens.

EreMaijala committed:
Drop user and database conditionally in shutdown.

EreMaijala committed:
Fix a typo.

EreMaijala committed:
Stop Solr conditionally on shutdown.

EreMaijala committed:
Delete apache conf conditionally.

EreMaijala commented:
@demiankatz I can reproduce by running shutdown without startup first, apart from the last `Class "Phing\Io\IOException" not found`, which I assume would be the result of the vendor directory going missing from under it. I believe the behavior is the same as before, it's just that Phing now emits the scary-looking errors about exec statuses (and looking at the source there's no way to suppress them).

I've pushed a set of tweaks and checks to alleviate the issue.

demiankatz commented:
Thanks, @EreMaijala. I think your theory about why I'm getting the missing class message makes sense, though it's strange that you're not getting the same thing.

Unfortunately, even with these changes, I'm still getting left in a bad state when I shutdown at the wrong time, which then breaks my ability to start up again (I think it bails out before the git reset, so then a bunch of files are missing).

Do you think perhaps it is time to stop deleting the vendor directory? I believe that we've talked about this in the past -- it's not necessarily helpful anyway, as long as startup runs composer to update things properly. It would save a lot of time if we didn't have to keep uninstalling and reinstalling all the dependencies... and if it's now breaking Phing, that's an excuse to make the change.

On a side note, I also notice that you pinned Composer at 2.7.1; is there a particular reason for that, given that newer versions are available?

EreMaijala committed:
Stop deleting vendor directory.

EreMaijala committed:
Bump composer version to 2.7.7.

EreMaijala committed:
Run composer install on startup.

EreMaijala commented:
@demiankatz Composer version was a mistake, now pinned to latest. I've removed deletion of vendor and added `composer install` to startup. Does this work any better for you?

demiankatz commented on build.xml:
```diff
@@ -457,28 +546,50 @@
 
+    <!-- Install / update dependencies -->
+    <phingcall target="composer" />
+    <exec executable="${srcdir}/composer.phar">
```
Isn't the composer target already calling composer install? Do we need to explicitly call it again here?

I also wonder if we should be downloading composer.phar into ${srcdir}/downloads so it's not cluttering up the root of the project... though that's probably just scope creep at this point and is better addressed in a follow-up. :-)

EreMaijala replied:
Of course, my mistake.

I decided to move composer.phar to downloads now while at it. :)

demiankatz replied:
Yeah, that was a weirder mess than I had realized; worth sorting out!

demiankatz reviewed (changes requested):
Thanks, @EreMaijala -- this seems to be working better for me, but see below for one more question.

I think whatever we do, there may be a transition period where we have to get used to the new version, and where we'll run into annoying problems when switching back and forth between branches. Probably best to just dive in and see how it goes... but let me know if you prefer to do that before or after the 10.0 release. :-)

EreMaijala committed:
Remove useless extra composer call.

EreMaijala committed:
Move composer to local directory.

EreMaijala committed:
Use downloads cache for pinned composer versions.

EreMaijala committed:
Keep composer in downloads directory.

EreMaijala commented:
@demiankatz I'd like to get this into 10.0 for proper PHP 8.3 support, but if you feel it's too risky, 10.1 would be fine too.

demiankatz reviewed (changes requested):
I've found a new problem that seems to be unique to Phing 3 -- it doesn't seem to have the same permissions as Phing 2 when deleting the cache. If I run code on the Doctrine branch in #2233, it creates a directory owned by Apache under $VUFIND_LOCAL_DIR/cache/configs to hold Doctrine-generated content (which all has 644 permissions). Running shutdown with Phing 2 clears this out with no problem, but doing it with this branch breaks the process with this message:

```
Buildfile: /usr/local/vufind/build.xml

vufind > shutdown:

   [delete] Deleting: /etc/apache2/conf-enabled/vufindtest.conf
     [exec] Result: 127
Sending stop command to Solr running on port 8983 ... waiting up to 180 seconds to allow Jetty process 19593 to stop gracefully.
     [exec] Result: 127
   [delete] Deleting: /usr/local/vufind/env.bat
   [delete] Deleting: /usr/local/vufind/env.sh
   [delete] Deleting: /usr/local/vufind/import/solrmarc.log
   [delete] Deleting directory /usr/local/vufind/solr
 [filesize] /usr/local/vufind/vufind-exception.log filesize is 0B
   [delete] Deleting: /usr/local/vufind/vufind-exception.log
   [delete] Deleting: /usr/local/vufind/vufind-mail.log
   [delete] Deleting directory /usr/local/vufind/local

BUILD FAILED
/usr/local/vufind/build.xml:1049:78 Unable to delete file /usr/local/vufind/local/cache/configs/b41555749e205d1b610f6a8eaed1702e/module-classmap-cache.php: Cannot delete /usr/local/vufind/local/cache/configs/b41555749e205d1b610f6a8eaed1702e/module-classmap-cache.php


Total time: 3.1199 seconds
```

Here's a directory listing of the offending cache:

```
dkatz@dkatz-VirtualBox:/usr/local/vufind$ ll /usr/local/vufind/local/cache/configs/b41555749e205d1b610f6a8eaed1702e/
total 352
drwxr-xr-x 2 www-data www-data   4096 Jun 18 14:39 ./
drwxr-xr-x 3 www-data www-data   4096 Jun 18 14:39 ../
-rw-r--r-- 1 www-data www-data    405 Jun 18 14:39 module-classmap-cache.php
-rw-r--r-- 1 www-data www-data 346745 Jun 18 14:39 module-config-cache.php
```

I imagine you can reproduce this problem by creating, chown-ing and chmod-ing some files without having to actually set up the Doctrine stuff. Any idea what's going on?

If we can get this sorted out, I'm not opposed to merging this close to the release date, since Phing is not normally used as part of the package unless people are pulling code from Git, so it's unlikely to break things for anyone even if there are some lingering issues that need to be addressed.

EreMaijala committed:
Add downloadsdir property.

EreMaijala committed:
Fix evaluation of non-boolean values.

EreMaijala committed:
Add messages

EreMaijala committed:
Fix execution of extra commands.

EreMaijala committed:
Add passthru.

EreMaijala commented:
@demiankatz Permission handling of mkdir has changed, but since the cache is not created by phing, I suspect the problem was that the extra_shutdown_cleanup task (or extra_startup_command) wasn't executed properly. If this is the case, it should be fixed now. I also changed the way e.g. `${apachectl}` is checked since boolean evaluation has changed and I'm not sure if it affects `<istrue>` as well.

EreMaijala committed:
Execute apachectl via shell.

EreMaijala committed:
Clean up white space.

EreMaijala committed:
Run phpunit via shell, fix target "phpunit".

demiankatz reviewed (approved):
Thanks, @EreMaijala!
